### PR TITLE
cache reference dialog into a dictionary when loading dup dialogs

### LINKED ISSUE #4491: PVA DCR: Object graph is too large for co-tenanted environments
# Problem
The declarative type loader creates a clone of an object on every reference.  So every place we refer to foo.dialog we load another copy of foo and stitch it into the object graph. PVA is worried about memory footprint for lots of co-hosted bots.

# Suggested solution
We should maintain a dictionary cache for the lifetime of loading the dialog references, so that every reference to foo.dialog simply is a pointer to the single instance of foo.dialog.

### PR DESCRIPTION
**Pull request by feich-ms**
Cache reference dialog into a dictionary when loading dup dialogs to make sure there is only one instance for same dialog reference in memory.

Fixes #4491

## Description
use dictionary caches to store duplicated reference dialog when loading dialog files to make sure there is only one instance object in memory for the same dialog.

## Testing
Test added

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs [hunk 1]
```diff
@@ -67,10 +70,20 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
 
             using (new SourceScope(sourceContext, range))
             {
+                string refDialogName = null;
                 if (this.resourceExplorer.IsRef(jToken))
                 {
-                    // We can't do this asynchronously as the Json.NET interface is synchronous
-                    jToken = this.resourceExplorer.ResolveRefAsync(jToken, sourceContext).GetAwaiter().GetResult();
+                    refDialogName = jToken.Value<string>();
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs:76]: Maybe should use "GetRefTarget()" ? But it seems a private method.

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs [hunk 2]
```diff
@@ -24,6 +24,9 @@ public class InterfaceConverter<T> : JsonConverter, IObservableConverter, IObser
         private readonly List<IJsonLoadObserver> observers = new List<IJsonLoadObserver>();
         private readonly SourceContext sourceContext;
 
+        private Dictionary<string, JToken> cachedRefJsons = new Dictionary<string, JToken>();
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs:27]: Use `ConcurrentDictionary` to ensure the thread safety

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs [hunk 3]
```diff
@@ -101,13 +114,28 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+                            cachedRefDialogs.Add(refDialogName, result);
+                        }
+                    }
 
                     // Associate the most specific source context information with this item
                     if (sourceContext.CallStack.Count > 0)
                     {
                         range = sourceContext.CallStack.Peek().DeepClone();
-                        DebugSupport.SourceMap.Add(result, range);
+                        if ((DebugSupport.SourceMap as Dictionary<object, SourceRange>).ContainsKey(result))
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs:135]: Why not use the "TryGetValue" in SourceMap,
```
if(!DebugSupport.SourceMap.TryGetValue(result, out var _))
{
   DebugSupport.SourceMap.Add(result, range);
}
```

Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs]: good suggestion. fixed.

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs (hunk 2 above)
Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs]: fixed.

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs (hunk 1 above)
Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs:77]: ResolveRefAsync is the public interface for resolving reference. I didn't add this, just removed it to the codnition when not cached.

**Comment by boydc2014** [comment]
@tomlm @carlosscastro  can you help take a look?

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs [hunk 4]
```diff
@@ -24,6 +25,9 @@ public class InterfaceConverter<T> : JsonConverter, IObservableConverter, IObser
         private readonly List<IJsonLoadObserver> observers = new List<IJsonLoadObserver>();
         private readonly SourceContext sourceContext;
 
+        private readonly ConcurrentDictionary<string, JToken> cachedRefJsons = new ConcurrentDictionary<string, JToken>();
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs:28]: >JToken [](start = 54, length = 6)

I think the intent was to cache the typed object, not the JToken.  Doing the JToken means we will be creating new instance on every reference as we convert from JTokn= > Typed object.

**Review by tomlm — CHANGES_REQUESTED** [review]
:clock1:

File: libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs (hunk 4 above)
Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Converters/InterfaceConverter.cs]: resovled

**Comment by feich-ms** [comment]
@tomlm comments resolved.

**Review by tomlm — APPROVED** [review]
:shipit:

**Pull request merged**